test_agent/rust/src/testagent.rs

```diff
@@ -0,0 +1,356 @@
+impl UListener for ListenerHandlers {
+    async fn on_receive(&self, msg: UMessage) {
+        dbg!("OnReceive called");
+
+        let data_payload = match &msg.payload.data {
+            Some(data) => {
+                // Now we have access to the Data enum
+                match data {
+                    Data::Reference(reference) => reference.to_string(),
+                    Data::Value(value) => {
+                        let value_str = String::from_utf8_lossy(value);
+                        value_str.to_string()
+                    }
+                    _ => "none".into(),
+                }
+            }
+            None => {
+                dbg!("No data available");
+                "none".into()
+            }
+        };
+
+        let Ok(value_str) = serde_json::to_string(
+            &[("value".to_string(), data_payload.to_string())]
+                .iter()
+                .cloned()
+                .collect::<HashMap<_, _>>(),
+        ) else {
+            error!("Issue in converting to payload");
+            return;
+        };
+
+        let Ok(payload_str) = serde_json::to_string(
+            &[("payload".to_string(), value_str.to_string())]
+                .iter()
+                .cloned()
+                .collect::<HashMap<_, _>>(),
+        ) else {
+            error!("Issue in converting to payload");
+            return;
+        };
+
+        let data = [("data".to_string(), payload_str.to_string())]
+            .iter()
+            .cloned()
+            .collect::<HashMap<_, _>>();
+        let json_message = JsonResponseData {
+            action: constants::RESPONSE_ON_RECEIVE.to_owned(),
+            data,
+            ue: "rust".to_string(),
+            test_id: "1".to_string(),
+        };
+
+        dbg!("sending received data to tm....");
+        let json_message_str = convert_json_to_jsonstring(&json_message);
+        let message = json_message_str.as_bytes();
+
+        let Ok(mut socket) = self.clientsocket_to_tm.try_lock() else {
+            error!("Failed to acquire lock for sending data to TM from on_receive");
+            return;
+        };
+
+        let result = socket.write_all(message);
+        match result {
+            Ok(()) => println!("on receive could send data to TM"),
+            Err(err) => error!("on receive could not send data to TM{}", err),
+        }
+    }
+
+    async fn on_error(&self, _err: UStatus) {
+        dbg!(_err);
+    }
+}
+
+impl SocketTestAgent {
+    pub fn new(test_clientsocket: TcpStream, listener: Arc<dyn UListener>) -> Self {
+        let clientsocket = Arc::new(Mutex::new(test_clientsocket));
+
+        Self {
+            clientsocket,
+            listener,
+        }
+    }
+
+    async fn handle_send_command(
+        &self,
+        utransport: &UTransportSocket,
+        json_data_value: Value,
+    ) -> Result<(), UStatus> {
+        let wrapper_umessage: WrapperUMessage = match serde_json::from_value(json_data_value) {
```

Comment:
I think the way you structured this is totally fine.

I'm a fan of early return too.

An alternate structure to this would have you do something like:
```rust
match serde_json::from_value(json_data_value) {
  Ok(message) => {
    utransport.send(message.0).await
  }
  Err(err) => {
    let err = format!("not able to deserialize send UMessage from Json{}", err);
    error!(err);
    Err(UStatus::fail_with_code(
      UCode::INTERNAL,
      err,
    ))
  }
}
```
But this may be a preference thing.

Reply:
maybe let's do this tmrw :)